Add support for File system listener.

harshaldev27 opened the pull request:
This Pull Request adds support for a File system listener, which allows QTEE to request both normal and secure file system services from REE.

The File system listener has a hard dependency on the "persist" partition, which is a separate storage partition available on Qualcomm platforms whose contents are preserved across Factory Reset and Over-the-Air update scenarios. This partition is used for storing encrypted files via the Secure File System (SFS) (also called the Global Platform File System (GPFS)) listener.

The File system listener also has a dependency on a Replay Protected Memory Block (RPMB) partition, for establishing a Root of Trust (RoT) for the SFS listener. However, in the absence of an RPMB partition, the OEM can move to an alternate (and less secure) RoT established via a version table (vTable) file created in the persist partition by the listener. This is not recommended on production devices.

This Pull Request has been tested on Qualcomm SM8650 platform via the `sampleapp` Trusted Application and a vTable Root of Trust.

quicbasantk commented on listeners/libdrmfs/msg.h:
```diff
@@ -0,0 +1,787 @@
+// Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+
+#ifndef __FS_MSG_H_
```
can we split this header into two part? fs and gpfs?

harshaldev27 replied:
Ack.

quicbasantk commented on listeners/libdrmfs/main.c:
```diff
@@ -0,0 +1,265 @@
+
+	/* Read command id */
+	switch (gpfs_cmd_id) {
+	case (TZ_GPFS_MSG_CMD_FILE_READ):
```
TZ_GPFS_MSG_CMD_FILE_READ/WRITE/RENAME/REMOVE cmd support should moved to depricated as QTEE has moved long from absolute path to related path.
Only keep and DATA and Persist cmd.

quicbasantk commented on listeners/libdrmfs/main.c:
```diff
@@ -0,0 +1,265 @@
+// Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+
```
This file can be split into two part?
Any reason we want to keep as sigle module?

harshaldev27 replied:
Ack.

quicbasantk commented on listeners/libdrmfs/helper.h:
```diff
@@ -0,0 +1,70 @@
+ * @param new_len Length of the new path buffer.
+ * @return const char* The modified path if applicable, otherwise the original.
+ */
+char *get_vendor_path(char *old_path, size_t old_len,
```
we need to find some better name, vendor keyword should be removed as this is influenced from Android?

harshaldev27 replied:
Ack. Maybe let's use `get_resolved_path()`.

quicbasantk commented on listeners/libdrmfs/gpfs.c:
```diff
@@ -0,0 +1,914 @@
+
+	if (rsp_len < sizeof(tz_gpfile_version_rsp_t)) {
+		MSGE("%s Invalid buffer length.\n", __func__);
+		return -1;
```
can we return some error which can signify failure like EINVAL or any lenght related error?

harshaldev27 replied:
Replied below.

quicbasantk commented on listeners/libdrmfs/gpfs.c:
```diff
@@ -0,0 +1,914 @@
+int gpfile_partition_error(void *rsp)
+{
+	tz_gpfile_err_rsp_t *my_rsp = (tz_gpfile_err_rsp_t *)rsp;
+
```
len check missing for error API?

harshaldev27 replied:
Ack.

quicbasantk commented on listeners/libdrmfs/gpfs.c:
```diff
@@ -0,0 +1,914 @@
+
+int gpfile_error(void *rsp)
+{
+	tz_gpfile_err_rsp_t *my_rsp = (tz_gpfile_err_rsp_t *)rsp;
```
len check missing for error API?

quicbasantk commented on listeners/libdrmfs/gpfs.c:
```diff
@@ -0,0 +1,914 @@
+	if (req_len < sizeof(tz_gpfile_rename_req_t) ||
+	    rsp_len < sizeof(tz_gpfile_rename_rsp_t)) {
+		MSGE("%s Invalid buffer length\n", __func__);
+		return -1;
```
can we return EINVAL or some nearby error code?
-1 is generic and hard to debug?

quicbasantk replied:
probably we should introdue new interface error code listener CBO and ensure invalid lenght error is getting passed to QTEE?
It will be help to improve debugging?

harshaldev27 replied:
This error code is used in file CListenerCBO.c. If we bubble an error from here, CListenerCBO.c interprets it as a transport error, and return `Object_ERROR` to QTEE. If you feel generic `Object_ERROR` is not good enough, we can try `Object_ERROR_INVALID`?

quicbasantk replied:
no, I mean error should be specific to CListenerCBO interface, OBJECT_ERROR_XXX is bit generic but we have reached CListenerCBO domain then we got error?

quicbasantk commented on listeners/libdrmfs/cmn.h:
```diff
@@ -0,0 +1,41 @@
+#define DATA_VENDOR_PATH        "/data/vendor"
+#define LEGACY_DATA_PATH        "/data/misc/qsee/"
+#define LEGACY_PERSIST_PATH     "/persist/data/"
+#define DATA_PATH               "/var/tmp/qtee_supplicant/vendor/tzstorage/"
```
tmp can be cleanup?
any better partition we can identify having read/write partition?

harshaldev27 replied:
Sorry, can you elaborate further?

quicbasantk replied:
I mean /var/tmp/qtee_supplicant --> can be erased without any explicit root permission?

harshaldev27 replied:
Qualcomm Linux is already using this partition path, changing this suddenly could lead to issues. I do not believe that random users can delete the contents of /var/tmp/qtee_supplicant, since this path is protected by SE policies and only qtee_supplicant daemon can access/modify the files under this path.

quicbasantk commented on libminkadaptor/tests/smcinvoke_client/src/smcinvoke_client.cpp:
```diff
@@ -552,6 +555,123 @@ static int loadApp(Object appLoader, std::string const &path,
+	switch (cmdId) {
+	case CLIENT_CMD5_RUN_GPFS_TEST:
+	case CLIENT_CMD6_RUN_FS_TEST:
+		ret = IAppLegacyTest_handleRequest(appLegacy, req, reqLen, &smplap_rsp, sizeof(smplap_rsp), &rspSizeOut);
```
This is tricky decision, do we want to plug dependency on sampleapp TA which we are planning to depricate from QLI perspective.
I think we should add one test case in QTEE SDK reference TA and use our test exposed SFS interface which we used in reset/recovery FR?

harshaldev27 replied:
We can continue to keep this test case for internal testing. And add the additional test case in `tzecotestapp.mbn`?

quicbasantk replied:
if we are thinking from reference and bare minimum unit perspective then we need some in QTEE SDK test TA, 
for unit and stress testing tzecotestap.mbn is way forward.

harshaldev27 replied:
We can develop a test case around QTEE SDK Test TA once the TA is released then.

harshaldev27 commented on qtee_supplicant/src/listener_mngr.c:
```diff
@@ -47,6 +46,34 @@ static struct listener_svc listeners[] = {
 		.buf_len = 0,
 	},
 #endif
+#ifdef FS_LISTENER
```
Ack.

quicbasantk commented on qtee_supplicant/src/CListenerCBO.c:
```diff
@@ -112,8 +112,8 @@ static int32_t CListenerCBO_request(ListenerCBO* me,
 
```
we will keep dlopen but only for init API, I mean dispatcher API is getting called from CListenerCBO interface which plug use case specific operation in that listener service.

I didn't get how ifdef cluttering will come into picture?

harshaldev27 replied:
As discussed offline, we will plan another commit which requires all listeners to absorb CListenerCBO.c in it's implementation and have a single callback definition in the listeners called init() which will be used to both register the listener and perform any one-time initialization. This will help will long term goal of deprecating the CListenerCBO callback and migrate to a context-based listener from a system wide listener.

harshaldev27 commented:
This patchset includes the following changes as discussed:
1. Split FS and GPFS listener implementation into separate files controlled by separate build flags.
2. Remove deprecated TZ_GPFS_MSG_CMD_FILE* command ids.
3. Re-named get_vendor_path() to get_resolved_path().
4. Added length checks for error returning functions.
5. Removed un-used init() callback definition from qtee_supplicant.
